web/packages/teleport/src/Roles/RoleEditor/RoleEditor.test.tsx

```diff
@@ -0,0 +1,294 @@
+/**
+ * Teleport
+ * Copyright (C) 2024 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+import { render, screen, userEvent } from 'design/utils/testing';
+
+import { within } from '@testing-library/react';
+
+import { UserEvent } from '@testing-library/user-event';
+
+import { Role } from 'teleport/services/resources';
+import { createTeleportContext } from 'teleport/mocks/contexts';
+import TeleportContextProvider from 'teleport/TeleportContextProvider';
+import { yamlService } from 'teleport/services/yaml';
+import {
+  YamlStringifyRequest,
+  YamlSupportedResourceKind,
+} from 'teleport/services/yaml/types';
+import { CaptureEvent, userEventService } from 'teleport/services/userEvent';
+
+import { RoleEditor, RoleEditorProps } from './RoleEditor';
+import { defaultOptions, withDefaults } from './withDefaults';
+
+// The Ace editor is very difficult to deal with in tests, especially that for
+// handling its state, we are using input event, which is asynchronous. Thus,
+// for testing, we just mock it out with a simple text area.
```

Comment:
Hmmmm, I didn't think about using `waitFor`. I'll try it out.

Reply:
I tried, but just can't. The Ace Editor throws every kind of obstacles at me, and it's not just asynchronicity. Right now, it just replaces everything I type with "XXXXXXXXXXXXXXXXXX". I'm losing my patience with this. I'm afraid it either goes in like this, or I remove the test if that's a problem. Though I'd really like to avoid this.

The best thing I can do is to maybe replace mocking Ace with mocking the TextEditor component, which I'm using directly. Would that be OK?

Reply:
Sorry you are having trouble with that. For your reference, here is one component we have for the SAML entity descriptor editor https://github.com/gravitational/teleport.e/blob/master/web/teleport/src/SamlApplication/components/EntityDescriptorEditor.tsx#L31-L68 and this is the test for that https://github.com/gravitational/teleport.e/blob/master/web/teleport/src/SamlApplication/components/ConfigureServiceProvider.test.tsx#L278-L286.

The trick is to keep the element focused using `userEvents` and `waitFor` to assert asynchronous changes. 

Feel free to skip this for now. Me personally, I would also remove the mocked tests as well as that gives false sense of assurance, unless you want to test a whole user interfaction where mocking might get you to test just the data flow and not the user interaction.